### PR TITLE
Add `neval_hprod` incrementation

### PR DESCRIPTION
**Pull request by paraynaud**

### CODE REVIEW

**Comment by paraynaud** [comment]
Following #86

File: test/nlp/quasi-newton.jl [hunk 1]
```diff
@@ -37,6 +37,7 @@
       @test obj(nlp, x) ≈ f(x)
       @test grad(nlp, x) ≈ ∇f(x)
       @test hprod(nlp, x, v) ≈ H(x) * v
+      @test nlp.model.counters.neval_hprod == 1
```
Comment (tmigot) [inline test/nlp/quasi-newton.jl:40]:
```suggestion
      @test neval_hprod(nlp) == 1
```

**Review by tmigot — COMMENTED** [review]
@paraynaud The more I think of it the more it feels slightly strange than `nlp.model` counts `hprod` evaluation while no hessian-products was used.

Maybe, we should just add a separate `Counters` for the `QuasiNewtonModel`?
So that `neval_hprod(nlp)` would return 1, but `neval_hprod(nlp.model)=0`.

**Comment by tmigot** [comment]
I am not sure to fully understand. Let's try to clarify, we have a 
```
nlp = ADNLPModel(f, x0)
model = LBFGSSolver(nlp)
```
and we do
```
hprod!(model, x0)
```
With your PR currently
```
neval_hprod(nlp) == 1
neval_hprod(model) == 1
```
because the LBFGSModel counter is synced with the nlp (actually, `model.counters` doesn't exist).
I agree that `neval_hprod(model) == 1` as the "hessian" of  `model` was used.
However, I find `neval_hprod(nlp) == 1` slightly confusing because no hessian of `nlp` has been used.

I think it would be accurate to have 
```
neval_hprod(nlp) == 0 # because no hessian was used
neval_hprod(model) == 1
```

**Comment by paraynaud** [comment]
I may have been confusing with LBFGS, which is a LBFGSModel used in trunk and not the inverse LBFGS linesearch from JSOSolvers.

I use:
```julia
nlp = ADNLPModel(f, x0)
model = LBFGSModel(nlp)
ges = trunk(model)
```

```quote
because the LBFGSModel counter is synced with the nlp (actually, model.counters doesn't exist).
```
correct!

```quote
I agree that neval_hprod(model) == 1 as the "hessian" of model was used.
However, I find neval_hprod(nlp) == 1 slightly confusing because no hessian of nlp has been used.
```
I did not think about it this way and I get your point.
I am not sure how to distinguish the two `neval_hprod` methods since `counter` is paired to `nlp`.

A suggestion: maybe we can link the `neval_hprod` method to the `counter` from `model.op.counter` to return the number of `mul!` performed by `model.op`, and delete the incrementation of `nlp.counter.neval_hprod` I add.
By doing that, `neval_hprod(nlp)` will stay at `0` all the time (as long a no `hprod` are performed directly on `nlp`).
What do you think about it?

**Comment by tmigot** [comment]
Yes, that would work. If I understand correctly, you plan to do something like this:
```
neval_hprod(model::QuasiNewtonModel) = model.op.counter
```
?

Do not hesitate to add this as a unit test:
```
neval_hprod(nlp) == 0 # because no hessian was used
neval_hprod(model) == 1
```

**Comment by paraynaud** [comment]
@tmigot I gave up writing it as 
```julia
NLPModels.neval_hprod(nlp::QuasiNewtonModel) = nlp.op.nprod
```
so I wrote it as
```julia
NLPModels.neval_hprod(nlp::LBFGSModel) = nlp.op.nprod
NLPModels.neval_hprod(nlp::LSR1Model) = nlp.op.nprod
```
and it behaves accordingly to our discussion.
I made a slight distinction in the tests, because a LSR1Model makes a `prod` during the `push!` method, which is not the case of a LBFGSModel.

**Review by tmigot — APPROVED** [review]
I feared it was going a large PR, so good job for keeping it concise. Looks good for me, thanks!

**Pull request merged**